FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+  }
+
+  void setValidTag(size_t Index, bool Valid) {
+    buffer[Index].first = Valid ? Slot::VALID : Slot::INVALID;
+  }
+
+  // Returns a mask to set in AbridgedTagWord
+  uint8_t getValidMask() {
+    uint8_t Mask = 0;
+    for (size_t i = 0; i < buffer.size(); i++) {
+      const bool Valid = buffer[i].first == Slot::VALID;
+      if (Valid) {
+        Mask |= 1U << i;
+      }
+    }
+    return Mask;
+  }
+
+  // Returns a mask to set in AbridgedTagWord
+  uint8_t getInvalidMask() {
+    uint8_t Mask = 0;
+    for (size_t i = 0; i < buffer.size(); i++) {
+      const bool Valid = buffer[i].first == Slot::INVALID;
+      if (Valid) {
+        Mask |= 1U << i;
+      }
+    }
+    return Mask;
+  }
+
+private:
+  fextl::deque<std::pair<Slot, T>> buffer;
+  // Real top as an offset from stored top value (or the one at the beginning of the block)
+  // For example, if we start and push a value to our simulated stack, because we don't
+  // update top straight away the TopOffset is 1.
+  // If SlowPath is true, then TopOffset is always zero.
+  int8_t TopOffset = 0;
+};
+
+
+class X87StackOptimization final : public FEXCore::IR::Pass {
+public:
+  X87StackOptimization() {
+    FEX_CONFIG_OPT(ReducedPrecision, X87REDUCEDPRECISION);
+    ReducedPrecisionMode = ReducedPrecision;
+  }
+  void Run(IREmitter* Emit) override;
+
+private:
+  bool ReducedPrecisionMode;
+  // FIXME(pmatos): copy from OpcodeDispatcher.h
+  [[nodiscard]]
+  uint32_t MMBaseOffset() {
+    return static_cast<uint32_t>(offsetof(Core::CPUState, mm[0][0]));
+  }
+  std::tuple<Ref, Ref> SplitF64SigExp(Ref Node);
+
+  // Top Management Helpers
+  /// Set the valid tag for Value as valid (if Valid is true), or invalid (if Valid is false).
+  void SetX87ValidTag(Ref Value, bool Valid);
+  // Generates slow code to load/store a value from the top of the stack
+  Ref LoadStackValueAtTop_Slow();
+  void StoreStackValueAtTop_Slow(Ref Value, bool SetValid = true);
+  // Generates slow code to load/store a value from an offset from the top of the stack
+  Ref LoadStackValueAtOffset_Slow(uint8_t Offset);
+  void StoreStackValueAtOffset_Slow(uint8_t Offset, Ref Value, bool SetValid = true);
+  // Update Top value in slow path for a pop
+  void UpdateTop4Pop_Slow();
+  void UpdateTop4Push_Slow();
+  // Synchronizes the current simulated stack with the actual values.
+  // Returns a new value for Top, that's synchronized between the simulated stack
+  // and the actual FPU stack.
+  Ref SynchronizeStackValues();
+  // Moves us from the fast to the slow path if ShouldMigrate is true.
+  void MigrateToSlowPathIf(bool ShouldMigrate);
+  // Top Cache Management
+  Ref GetTopWithCache_Slow();
+  Ref GetOffsetTopWithCache_Slow(uint8_t Offset);
+  void SetTopWithCache_Slow(Ref Value);
+  Ref GetX87ValidTag_Slow(uint8_t Offset);
+  // Resets fields to initial values
+  void Reset(bool AlsoSlowPath = true);
+
+  struct StackMemberInfo {
+    IR::OpSize SourceDataSize; // Size of SourceDataNode
```

Comment:
You might need a `using IR` at the top of the file, but there should not be any need for `IR::` in the pass.

Reply:
Actually I didn't even need that cause the namespace is already `FEXCore::IR`.